### PR TITLE
Created test for onSelectedTags.

### PR DESCRIPTION
**Pull request by Haz3-jolt**
## Purpose / Description
Added Unit test to function with @NeedsTest.

## Fixes
* Created test for onSelectedTags to address "14656: adding tags does not flip the card", and "does not hang and can be called twice".

## Approach
- Unit test with multiple cases to test for edge cases.
## How Has This Been Tested?
- Ran the Unit test multiple times, and ensured it is passing.
- Verified that Logcat messages are logged.

## Checklist
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [x] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt [hunk 1]
```diff
@@ -137,6 +140,47 @@ class AbstractFlashcardViewerCommandTest : RobolectricTest() {
...
+
+            // Assert that the card is not flipped
+            assertFalse(viewer.isDisplayingAnswer)
+
+            Timber.d("Test", "Before second call to onSelectedTags")
+
+            // Call onSelectedTags method again
+            viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
+
+            Timber.d("Test", "After second call to onSelectedTags")
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt:177]: Is this how Timber works? I would expect it just to print "Test"

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt [hunk 2]
```diff
@@ -137,6 +140,47 @@ class AbstractFlashcardViewerCommandTest : RobolectricTest() {
...
+            // Create a list of tags
+            val tags = listOf("tag1", "tag2")
+
+            // Define an arbitrary filter
+            val ARBITRARY_FILTER = CardStateFilter.DUE
+
+            // Check if currentCard is null before calling onSelectedTags
+            if (viewer.currentCard == null) {
+                Timber.d("Test", "currentCard is null, not calling onSelectedTags")
+                return@use
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt:159]: This should fail the test

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt [hunk 3]
```diff
@@ -217,7 +261,7 @@ class AbstractFlashcardViewerCommandTest : RobolectricTest() {
         return c
     }
 
-    private class CommandTestCardViewer(private var currentCardOverride: Card?) : Reviewer() {
+    class CommandTestCardViewer(private var currentCardOverride: Card?) : Reviewer() {
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt:264]: revert - not necessary

**Review by david-allison — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt (hunk 1 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt:175]: The first parameter is the tag and the second parameter is the message. works for Log so should work here

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt (hunk 2 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt]: modified to assertNotNull, it was cleaner

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt (hunk 3 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt]: done, did it accidentally.

**Comment by david-allison** [comment]
```
junit.framework.AssertionFailedError: currentCard should not be null
	at junit.framework.Assert.fail(Assert.java:57)
	at junit.framework.Assert.assertTrue(Assert.java:22)
	at junit.framework.Assert.assertNotNull(Assert.java:256)
	at junit.framework.TestCase.assertNotNull(TestCase.java:399)
	at com.ichi2.anki.AbstractFlashcardViewerCommandTest.testOnSelectedTags(AbstractFlashcardViewerCommandTest.kt:158)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at com.ichi2.testutils.FailOnUnhandledExceptionRule$apply$1.evaluate(FailOnUnhandledExceptionRule.kt:57)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:489)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:290)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:104)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)
	Suppressed: org.robolectric.android.internal.AndroidTestEnvironment$UnExecutedRunnablesException: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(Looper.getMainLooper()).idle() call.
```

File: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt:175]: The first argument is the message, the second is the arg for the first `%s`

Screenshot 2024-04-30 at 21 50 12

**Comment by david-allison** [comment]
```patch
Index: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt
--- a/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt	(revision 42ef65b9672df208eb2a81f0c3f66e1867b484e8)
+++ b/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt	(date 1714510166175)
@@ -16,6 +16,7 @@
 package com.ichi2.anki
 
 import androidx.test.ext.junit.runners.AndroidJUnit4
+import com.ichi2.anki.AbstractFlashcardViewerTest.NonAbstractFlashcardViewer
 import com.ichi2.anki.cardviewer.ViewerCommand
 import com.ichi2.anki.cardviewer.ViewerCommand.*
 import com.ichi2.anki.cardviewer.ViewerRefresh
@@ -111,7 +112,7 @@
     @Test
     fun testRefreshIfRequired() {
         // Create an ActivityController instance for AbstractFlashcardViewer
-        Robolectric.buildActivity(AbstractFlashcardViewerTest.NonAbstractFlashcardViewer::class.java).use { controller ->
+        Robolectric.buildActivity(NonAbstractFlashcardViewer::class.java).use { controller ->
             // Case 1: Activity is resuming, refreshRequired is not null
             with(controller.create().start().get()) {
                 refreshRequired = updateCard
@@ -143,40 +144,37 @@
 
     @Test
     fun testOnSelectedTags() {
-        // Create an ActivityController instance for AbstractFlashcardViewer
-        Robolectric.buildActivity(AbstractFlashcardViewerTest.NonAbstractFlashcardViewer::class.java).use { controller ->
-            // Create and start the activity
-            val viewer = controller.get()
+        addNoteUsingBasicModel()
+        val viewer = startRegularActivity<Reviewer>()
 
-            // Create a list of tags
-            val tags = listOf("tag1", "tag2")
+        // Create a list of tags
+        val tags = listOf("tag1", "tag2")
 
-            // Define an arbitrary filter
-            val ARBITRARY_FILTER = CardStateFilter.DUE
+        // Define an arbitrary filter
+        val ARBITRARY_FILTER = CardStateFilter.DUE
 
-            // Assert that currentCard is not null before calling onSelectedTags
-            assertNotNull("currentCard should not be null", viewer.currentCard)
+        // Assert that currentCard is not null before calling onSelectedTags
+        assertNotNull("currentCard should not be null", viewer.currentCard)
 
-            Timber.d("Test", "Before first call to onSelectedTags")
+        Timber.d("Test", "Before first call to onSelectedTags")
 
-            // Call onSelectedTags method
-            viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
+        // Call onSelectedTags method
+        viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
 
-            Timber.d("Test", "After first call to onSelectedTags")
+        Timber.d("Test", "After first call to onSelectedTags")
 
-            // Assert that the card is not flipped
-            assertFalse(viewer.isDisplayingAnswer)
+        // Assert that the card is not flipped
+        assertFalse(viewer.isDisplayingAnswer)
 
-            Timber.d("Test", "Before second call to onSelectedTags")
+        Timber.d("Test", "Before second call to onSelectedTags")
 
-            // Call onSelectedTags method again
-            viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
+        // Call onSelectedTags method again
+        viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
 
-            Timber.d("Test", "After second call to onSelectedTags")
+        Timber.d("Test", "After second call to onSelectedTags")
 
-            // Assert that the card is not flipped
-            assertFalse(viewer.isDisplayingAnswer)
-        }
+        // Assert that the card is not flipped
+        assertFalse(viewer.isDisplayingAnswer)
     }
 
     @Test
```

**Comment by Haz3-jolt** [comment]
> ```
> junit.framework.AssertionFailedError: currentCard should not be null
> 	at junit.framework.Assert.fail(Assert.java:57)
> 	at junit.framework.Assert.assertTrue(Assert.java:22)
> 	at junit.framework.Assert.assertNotNull(Assert.java:256)
> 	at junit.framework.TestCase.assertNotNull(TestCase.java:399)
> 	at com.ichi2.anki.AbstractFlashcardViewerCommandTest.testOnSelectedTags(AbstractFlashcardViewerCommandTest.kt:158)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> 	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
> 	at com.ichi2.testutils.FailOnUnhandledExceptionRule$apply$1.evaluate(FailOnUnhandledExceptionRule.kt:57)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:489)
> 	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$2(SandboxTestRunner.java:290)
> 	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:104)
> 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
> 	at java.base/java.lang.Thread.run(Thread.java:840)
> 	Suppressed: org.robolectric.android.internal.AndroidTestEnvironment$UnExecutedRunnablesException: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(Looper.getMainLooper()).idle() call.
> ```

Yeah, accidentally sent it earlier version, fixed now.

**Comment by Haz3-jolt** [comment]
> In addition, this is in the wrong file, bring it to `ReviewerTest`
> 
> In addition: I suspect this won't fail before the commit was added, due to the tags dialog using `ChangeManager`.
> 
> But I may be wrong here and the test may be sufficient
> 
> ```diff
> Index: AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt
> IDEA additional info:
> Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
> <+>UTF-8
> ===================================================================
> diff --git a/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt b/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt
> --- a/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt	(revision 42ef65b9672df208eb2a81f0c3f66e1867b484e8)
> +++ b/AnkiDroid/src/test/java/com/ichi2/anki/AbstractFlashcardViewerCommandTest.kt	(date 1714510166175)
> @@ -16,6 +16,7 @@
>  package com.ichi2.anki
>  
>  import androidx.test.ext.junit.runners.AndroidJUnit4
> +import com.ichi2.anki.AbstractFlashcardViewerTest.NonAbstractFlashcardViewer
>  import com.ichi2.anki.cardviewer.ViewerCommand
>  import com.ichi2.anki.cardviewer.ViewerCommand.*
>  import com.ichi2.anki.cardviewer.ViewerRefresh
> @@ -111,7 +112,7 @@
>      @Test
>      fun testRefreshIfRequired() {
>          // Create an ActivityController instance for AbstractFlashcardViewer
> -        Robolectric.buildActivity(AbstractFlashcardViewerTest.NonAbstractFlashcardViewer::class.java).use { controller ->
> +        Robolectric.buildActivity(NonAbstractFlashcardViewer::class.java).use { controller ->
>              // Case 1: Activity is resuming, refreshRequired is not null
>              with(controller.create().start().get()) {
>                  refreshRequired = updateCard
> @@ -143,40 +144,37 @@
>  
>      @Test
>      fun testOnSelectedTags() {
> -        // Create an ActivityController instance for AbstractFlashcardViewer
> -        Robolectric.buildActivity(AbstractFlashcardViewerTest.NonAbstractFlashcardViewer::class.java).use { controller ->
> -            // Create and start the activity
> -            val viewer = controller.get()
> +        addNoteUsingBasicModel()
> +        val viewer = startRegularActivity<Reviewer>()
>  
> -            // Create a list of tags
> -            val tags = listOf("tag1", "tag2")
> +        // Create a list of tags
> +        val tags = listOf("tag1", "tag2")
>  
> -            // Define an arbitrary filter
> -            val ARBITRARY_FILTER = CardStateFilter.DUE
> +        // Define an arbitrary filter
> +        val ARBITRARY_FILTER = CardStateFilter.DUE
>  
> -            // Assert that currentCard is not null before calling onSelectedTags
> -            assertNotNull("currentCard should not be null", viewer.currentCard)
> +        // Assert that currentCard is not null before calling onSelectedTags
> +        assertNotNull("currentCard should not be null", viewer.currentCard)
>  
> -            Timber.d("Test", "Before first call to onSelectedTags")
> +        Timber.d("Test", "Before first call to onSelectedTags")
>  
> -            // Call onSelectedTags method
> -            viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
> +        // Call onSelectedTags method
> +        viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
>  
> -            Timber.d("Test", "After first call to onSelectedTags")
> +        Timber.d("Test", "After first call to onSelectedTags")
>  
> -            // Assert that the card is not flipped
> -            assertFalse(viewer.isDisplayingAnswer)
> +        // Assert that the card is not flipped
> +        assertFalse(viewer.isDisplayingAnswer)
>  
> -            Timber.d("Test", "Before second call to onSelectedTags")
> +        Timber.d("Test", "Before second call to onSelectedTags")
>  
> -            // Call onSelectedTags method again
> -            viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
> +        // Call onSelectedTags method again
> +        viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
>  
> -            Timber.d("Test", "After second call to onSelectedTags")
> +        Timber.d("Test", "After second call to onSelectedTags")
>  
> -            // Assert that the card is not flipped
> -            assertFalse(viewer.isDisplayingAnswer)
> -        }
> +        // Assert that the card is not flipped
> +        assertFalse(viewer.isDisplayingAnswer)
>      }
>  
>      @Test
> ```

Moved it, and yeah you were right about it not failing before then. should be fixed now though!

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt [hunk 4]
```diff
@@ -66,6 +70,46 @@ class ReviewerTest : RobolectricTest() {
...
+
+                // Call onSelectedTags method again
+                viewer.onSelectedTags(tags, emptyList(), ARBITRARY_FILTER)
+
+                Timber.d("Test", "After second call to onSelectedTags")
+
+                // Assert that the card is not flipped
+                TestCase.assertFalse(viewer.isDisplayingAnswer)
+            } ?: run {
+                Timber.d("Test", "currentCard is null, not calling onSelectedTags")
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:108]: If it's null, it should fail the test

Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:108]: oh ok!

**Comment by david-allison** [comment]
I'd advise looking back at the provided patch, as it fixes a few issues

**Comment by Haz3-jolt** [comment]
> I'd advise looking back at the provided patch, as it fixes a few issues

Okay, On it!

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt [hunk 5]
```diff
@@ -66,6 +69,44 @@ class ReviewerTest : RobolectricTest() {
...
+        // Create a list of tags
+        val tags = listOf("tag1", "tag2")
+
+        // Define an arbitrary filter
+        val ARBITRARY_FILTER = CardStateFilter.DUE
+
+        // Assert that currentCard is not null before calling onSelectedTags
+        assertNotNull("currentCard should not be null", viewer.currentCard)
+
+        Timber.d("Test", "Before first call to onSelectedTags")
```
Comment (david-allison) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:89]: Timber calls still need fixing, other than that, great!

**Review by david-allison — CHANGES_REQUESTED** [review]

**Comment by david-allison** [comment]
One final note, to confirm you tested and got a failure:

> In addition: I suspect this won't fail before the commit was added, due to the tags dialog using ChangeManager.
> 
> But I may be wrong here and the test may be sufficient

**Comment by Haz3-jolt** [comment]
> One final note, to confirm you tested and got a failure:
> 
> > In addition: I suspect this won't fail before the commit was added, due to the tags dialog using ChangeManager.
> > But I may be wrong here and the test may be sufficient

yep can confirm!

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt (hunk 5 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:89]: Almost Done!

**Review by david-allison — APPROVED** [review]
Perfect!

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt [hunk 6]
```diff
@@ -66,6 +69,44 @@ class ReviewerTest : RobolectricTest() {
...
+
+        // Create a list of tags
+        val tags = listOf("tag1", "tag2")
+
+        // Define an arbitrary filter
+        val ARBITRARY_FILTER = CardStateFilter.DUE
+
+        // Assert that currentCard is not null before calling onSelectedTags
+        assertNotNull("currentCard should not be null", viewer.currentCard)
+
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: perhaps verify all the expected state possible prior to the first action?
(note also that I have a slight preference for positively asserting the exact thing you want, vs negatively asserting one thing from a set of perhaps more than one thing you don't want... thus assert true on question vs assert false on answer - that said I have no idea if that method exists I am just typing this into a text box on github PR reviewer UI)

```suggestion
        assertTrue(viewer.isDisplayingQuestion)
```

Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: The method doesn't exist, but still I get what you mean. But i think here it should be fine. But if you think otherwise that is fine too, just hit me up and ill Change it.

Comment (david-allison) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: We typically use: `!viewer.isDisplayingAnswer`

Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: got it

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt [hunk 7]
```diff
@@ -66,6 +70,44 @@ class ReviewerTest : RobolectricTest() {
...
+        val viewer = startRegularActivity<Reviewer>()
+
+        // Create a list of tags
+        val tags = listOf("tag1", "tag2")
+
+        // Define an arbitrary filter
+        val ARBITRARY_FILTER = CardStateFilter.DUE
+
+        // Assert that currentCard is not null before calling onSelectedTags
+        assertNotNull("currentCard should not be null", viewer.currentCard)
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: My suggestion had two parts:
1- make an assertion against all possible state *prior* to changing state
2- make the assertion in an assertTrue style (if possible)

Per your and David's comments the method I just typed into the text box does not exist, but that's okay. Item 2 was a minor point

Item 1 is more important to me though. I think if you are going to assert on some state later (that viewer.isDisplayingAnswer has some value you expect), you had better assert that it has a known state prior to running the code under test. 

That's what I meant by asserting on all the the things you could.

Concretely then, with an API that apparently exists in a style that works (for legacy reasons) I mean add this assertion right here

```suggestion
        assertNotNull("currentCard should not be null", viewer.currentCard)
        assertTrue(!viewer.isDisplayingAnswer)
```

**Review by mikehardy — APPROVED** [review]
I do appreciate the test a lot - and with the added assertion, that much better

**Comment by david-allison** [comment]
Echoing Mike: you're doing an **AWESOME** job with tests

File: AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt (hunk 7 above)
Comment (Haz3-jolt) [reply AnkiDroid/src/test/java/com/ichi2/anki/ReviewerTest.kt:88]: Yeah about that, sorry mate I was traveling and couldn't add it in properly, was doing it on my phone remotely, and didn't read it properly 😅
